### ISSUE TITLE
Uncaught (in promise) TypeError: Cannot read property 'bufferId' of null

### ISSUE THREAD

**Issue by zertosh** [atom/atom#12280]
### Prerequisites

* [x] Can you reproduce the problem in [safe mode](http://flight-manual.atom.io/hacking-atom/sections/debugging/#using-safe-mode)?
* [x] Are you running the [latest version of Atom](http://flight-manual.atom.io/hacking-atom/sections/debugging/#update-to-the-latest-version)?
* [x] Did you check the [debugging guide](http://flight-manual.atom.io/hacking-atom/sections/debugging/)?
* [x] Did you check the [FAQs on Discuss](https://discuss.atom.io/c/faq)?
* [x] Are you reporting to the [correct repository](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#atom-and-packages)?
* [x] Did you [perform a cursory search](https://github.com/issues?q=is%3Aissue+user%3Aatom+-repo%3Aatom%2Felectron) to see if your bug or enhancement is already reported?

For more information on how to write a good [bug report](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#how-do-i-submit-a-good-bug-report) or [enhancement request](https://github.com/atom/atom/blob/master/CONTRIBUTING.md#how-do-i-submit-a-good-enhancement-suggestion), see the `CONTRIBUTING` guide.

### Description

`ctrl+alt+cmd+L` (or `cmd+R` with the dev tools open) with an unsaved editor will reload to an empty window because of:

```
Uncaught (in promise) TypeError: Cannot read property 'bufferId' of null
module.exports.TokenizedBuffer.deserialize  @ /Applications/Atom.app/Contents/Resources/app.asar/src/tokenized-buffer.js:51
module.exports.TextEditor.deserialize @ /Applications/Atom.app/Contents/Resources/app.asar/src/text-editor.js:98
module.exports.DeserializerManager.deserialize  @ /Applications/Atom.app/Contents/Resources/app.asar/src/deserializer-manager.js:45
(anonymous function)  @ /Applications/Atom.app/Contents/Resources/app.asar/src/pane.js:37
module.exports.Pane.deserialize @ /Applications/Atom.app/Contents/Resources/app.asar/src/pane.js:36
module.exports.DeserializerManager.deserialize  @ /Applications/Atom.app/Contents/Resources/app.asar/src/deserializer-manager.js:45
module.exports.PaneContainer.deserialize  @ /Applications/Atom.app/Contents/Resources/app.asar/src/pane-container.js:64
module.exports.Workspace.deserialize  @ /Applications/Atom.app/Contents/Resources/app.asar/src/workspace.js:174
module.exports.AtomEnvironment.deserialize  @ /Applications/Atom.app/Contents/Resources/app.asar/src/atom-environment.js:1031
(anonymous function)  @ /Applications/Atom.app/Contents/Resources/app.asar/src/atom-environment.js:752
```

### Steps to Reproduce

1. Using either Atom 1.9.0 or Atom 1.10.0-beta0 (in safe mode or not), have an unsaved editor and no project
2. Press `ctrl+opt+cmd+L`,
3. Reload to a blank screen (check the console for the error),
4. Press `ctrl+opt+cmd+L` to re-reload,
5. Works again.

**Expected behavior:** [What you expected to happen]

For the `ctrl+opt+cmd+L` reload to work fine the first time.

**Actual behavior:** [What actually happened]

Had to reload twice because the first time throws an error that leaves Atom in an unusable state.

### Versions

You can get this information from executing `atom --version` and `apm --version` at the command line. Also, please include the OS and what version of the OS you're running.

Atom 1.9.0 and Atom 1.10.0-beta0, on macOS 10.11.

**Comment by 50Wliu** [comment]
I can't reproduce this using Atom 1.10.0-dev-7e006cd, Windows 10 x64.

**Comment by zertosh** [comment]
I just tried building from master@ed5ad3c and still happens. I also removed `~/.atom` and `~/Library/Application Support/Atom`, and still happens.

**Comment by matthewwithanm** [comment]
I can repro this too with an unsaved editor and no project.

It looks like the issue is in the [TextEditor deserialization][1]:

```coffee
    if state.version isnt @prototype.serializationVersion and state.displayBuffer?
      state.tokenizedBuffer = state.displayBuffer.tokenizedBuffer
```

Both versions are `1` so `tokenizedBuffer` is never set. That means it passes `null` to the TokenizedBuffer deserializer.

Not sure what the intent of the version check is here…it seems like this could be something like `state.tokenizedBuffer or= state.displayBuffer?.tokenizedBuffer`?

[1]: https://github.com/atom/atom/blob/a7c38ea732489a02a3bd0207b466b4c66cfb4281/src/text-editor.coffee#L96-L97

**Comment by idleberg** [comment]
> I can repro this too with an unsaved editor and no project.

Same here. Also, this only happens when there hasn't been a non-editor (e.g. settings) pane open.

**Comment by lee-dohm** [comment]
I can only repro this with an unsaved editor and no project. I updated the repro steps to reflect that.

**Comment by 50Wliu** [comment]
Reproduced with the new steps on Windows.

**Comment by nathansobo** [comment]
Investigating...

**Comment by nathansobo** [comment]
I've discovered the issue. This one is super nuanced.

When the window is not associated with a project, we store the window state on the window itself rather than in the indexedDB, which is what we normally do. This is because our database keys are based on the folders loaded in the project, but we don't have a project in this case. (It would probably be better to use the same indexedDB code path and associate an id with the window, but this is how it's evolved).

Anyway, when we introduced display layers, we changed the serialization format of `TextEditor` to include a `tokenizedBuffer` key at the top level of the object instead of burying it inside of the `displayBuffer` key, because display buffers no longer exist. But for backward compatibility, we also retained a redundant copy of the `tokenizedBuffer` state under a `displayBuffer` key. We did this so that for a time, you can fall backward from 1.9 to 1.8 and still have editors serialize correctly.

Unfortunately, we actually [return the same *reference* to the tokenized buffer state in both places in the serialized object](https://github.com/atom/atom/blob/d2ed4712ab7b36f5a50d9fa13f40ecb025de04d3/src/text-editor.coffee#L191). This works fine when writing into IndexedDB, but apparently when serializing the object to send it back to the browser process, we're only allowed a single instance in the object tree and the other reference is assigned to `null`...

Try this in the console if you want to see for yourself:

```js
var x = {a: 1}
var y = {b: x, c: x}
atom.applicationDelegate.setTemporaryWindowState(y)
atom.applicationDelegate.getTemporaryWindowState().then(s => console.log(s))
// output: {b: {a: 1}, c: null}
```

/cc @maxbrunsfeld @as-cii

**Comment by maxbrunsfeld** [comment]
@nathansobo's fix will be released in 1.9.8 and 1.10.0-beta5.
